lib/block-supports/layout.php

```diff
@@ -291,6 +291,28 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
 	return '';
 }
 
+/**
+ * Indentifies inner block wrapper classnames.
+ *
+ * @param array $inner_content An array of strings.
+ * @return string String of inner wrapper classnames.
+ */
+function gutenberg_identify_inner_block_wrapper_classnames( $inner_content ) {
+
+	if ( ! isset( $inner_content[0] ) ) {
+		return false;
+	}
+
+	$inner_content_chunk      = new WP_HTML_Tag_Processor( $inner_content[0] );
+	$inner_wrapper_classnames = '';
+
+	while ( true === $inner_content_chunk->next_tag() ) {
+		$inner_wrapper_classnames = $inner_content_chunk->get_attribute( 'class' );
+	}
```

Comment:
a couple thoughts on this function:

we're passing in inner blocks but then trying to get the first one. the additional list-handling here might not be helping the clarity of its purpose. if we passed in _a block_ instead of a list of blocks, or instead of _inner content_, then it could focus on a more specific purpose.

```php
function get_classes_from_last_tag( $html ) {
	$tags         = new WP_HTML_Tag_Processor( $html );
	$last_classes = '';

	while ( $tags->next_tag() ) {
		$last_classes = $tags->get_attribute( 'class' ) || '';
	}

	return $last_classes;
}
```

Secondly, while I think the code as-is is fine, we don't need the `true ===` in the `while` condition because as a boolean value that's already implicit. No problem keeping it in there, but this isn't like functions which return a _truthy_ value; it actually returns `true` or `false`.

Third, `get_attribute( 'class' )` might return `null` if the last tag doesn't have a `class` attribute. The return type in the docblock comment says this function returns a `string` but unless we add a fallback it will also return `null` from time to time.

Fourth, a block's `$inner_content` is not guaranteed to contain only strings. It may contain `null` values where inner blocks belong. Another benefit to focusing this function on HTML processing and letting the calling code extract the HTML it wants to scan is not having to deal with this confusion here.

```php
$inner_content_classnames = isset( $block['innerContent'][0] ) && 'string' === typeof $block['innerContent'][0]
	? get_classes_from_last_tag( $block['innerContent'][0] )
	: '';
```

@adamziel I wonder; if we exposed the offset/position in the stream we could allow people to add their own "rewind" tracking and other more complicated parsing. would that be worth the risk it exposes?

```php
$last_tag_at = null;
while ( $tags->next_tag() ) {
	$last_tag_at = $tags->current_position();
}

if ( null === $last_tag_at ) {
	return;
}

$tags->seek( $last_tag_at );
return $tags->get_attribute( 'class' ) || '';
```

this could expose a lot of functionality it seems people want to do with this, as long as we don't expose the internal pointer, such that the offset points to an old version of the document, before it was changed by `set_attribute()` calls… 🤔

Reply:
Ooh, that's great feedback, thanks! Yes, the function is a lot neater in your example 😄

Reply:
> @adamziel I wonder; if we exposed the offset/position in the stream we could allow people to add their own "rewind" tracking and other more complicated parsing. would that be worth the risk it exposes?
 
Hm hm hm. This worries me in the same way as exposing `get_classes()` [worried you in the other thread](https://github.com/WordPress/gutenberg/pull/44600#discussion_r997552055). The cursor is an implementation detail and I'd rather not have people mess around with it – it would make their core fragile and severely constraint the future updates to the tag processor. For example, `seek` opens the following can of worms:

```php
$tags->next_tag();
$tag_1_at = $tags->current_position();

$tags->next_tag();
$tag_2_at = $tags->current_position();

$tags->seek( $tag_1_at );
$tags->set_attribute( 'class', 'bold' );

// $tag_2_at no longer points where we expect it to
$tags->seek( $tag_2_at );
$tags->set_attribute( 'class', 'bold' );
```

I'd rather keep an eye out for the use-cases that come up and see if we can address most of them with case-specific APIs. For example, `has_next_tag` would turn this loop into:

````php
while($tags->has_next_tag()) {
    $tags->next_tag();
}
return $tags->get_attribute( 'class' );
```